discovery: unify rate.Limiter across all gossip peers

Roasbeef opened the pull request:
In this commit, we revamp the old message based rate limiting. First, we
move to meter by bytes/s instead of messages/s. The old logic had an
error in that it limited groups of message replies, instead of each
message. With this new approach, we'll use the newly added
SerializedSize method to implement fine grained bandwidth metering.

We need to pick two values, the burst rate, and the msg bytes rate. The
burst rate is the max amt that can be sent in a given period of time. We
need to set this above 65 KB, or the max msg limit, otherwise no
messages can be sent. The bucket starts with this many tokens (bytes).
As those are depleted, the amount of tokens is refilled at the msg
bytes rate.

As conservative values, we've chosen 200 KB as the burst rate, and 100
KB/s as the limit.

Depends on https://github.com/lightningnetwork/lnd/pull/9623

Roasbeef commented on discovery/syncer.go:
```diff
@@ -1378,6 +1356,11 @@ func (g *GossipSyncer) ApplyGossipFilter(filter *lnwire.GossipTimestampRange) er
 		defer returnSema()
 
 		for _, msg := range newUpdatestoSend {
+			if err := g.maybeAddMsgDelay(); err != nil {
```
Looking at the diff again, I think ti'll just move this call into `sendToPeerSync`.

morehouse commented:
Concept ACK.

This makes sense -- the rate limit is there to throttle outgoing bandwidth, so it should apply globally rather than per-peer.

Roasbeef commented:
Pushed up an updated variant where we'll use `rate.Limiter` to implement a global outbound bandwidth rate limiter. We'll likely need to tune the defaults somewhat.

yyforyongyu reviewed (commented):
Looks like the changes got rebased out😂

Roasbeef commented:
> Looks like the changes got rebased out😂

LOL, not sure what happened there...

FIxed!

yyforyongyu commented on discovery/sync_manager_test.go:
```diff
@@ -685,3 +689,174 @@ func assertActiveSyncerTransition(t *testing.T, s *GossipSyncer, peer *mockPeer)
+
+// TestMaybeAddMsgDelay tests that the SyncManager.maybeAddMsgDelay method
+// correctly computes delays based on message size.
+func TestMaybeAddMsgDelay(t *testing.T) {
```
👍

yyforyongyu commented on discovery/syncer.go:
```diff
@@ -154,7 +155,7 @@ func (s syncerState) String() string {
 	// DefaultMaxUndelayedQueryReplies specifies how many gossip queries we
 	// will respond to immediately before starting to delay responses.
-	DefaultMaxUndelayedQueryReplies = 10
+	DefaultMaxUndelayedQueryReplies = 50
```
This can be removed now

Roasbeef replied:
Done. 

Part of me wonders if we should keep a base line msg/s second limit _in addition_ to the bandwidth limit.

Roasbeef replied:
Just thinking about a degenerate case where someone is having us send super small messages, so while we're under the total limit, we're doing more work overall.

Crypt-iQ replied:
I think this only matters for messages we send out not ones that they send us, so the size of our sent messages is _less_ under their control (still under their control somewhat since they are querying).

Crypt-iQ replied:
I think it'd be wise to also limit the number of messages so that they don't potentially sit in memory in the case of routing dumps, but this _should_ be addressed by the existing syncer semaphore.

yyforyongyu commented on discovery/sync_manager.go:
```diff
@@ -494,6 +543,94 @@ func (m *SyncManager) isPinnedSyncer(s *GossipSyncer) bool {
+func (m *SyncManager) maybeRateLimitMsg(ctx context.Context, peerPub [33]byte,
+	msg lnwire.Message) error {
+
+	delay, err := m.maybeAddMsgDelay(ctx, peerPub, msg)
```
It looks like we can just call `m.rateLimiter.Wait` here, which combines the two methods `maybeAddMsgDelay` and `waitMsgDelay`.

Roasbeef replied:
So `Wait` will only attempt to consume a single token, a token here is a single byte. `WaitN` is similar to `ReserveN`, but it'll actually _error out_ if the rate limit is exceeded. Rather than error out and drop _our_ send, I think we want to just wait until more room is available in the bucket.

Calling `ReserveN` instead is nice as it lets us log the delay (it returns the delay for you to sleep yourself instead of the library sleeping for you). 

I split it up into two methods so we can log the delay, and we can also use `maybeAddDelayMsg` for unit tests, as it returns the delay instead of sleeping itself.

yyforyongyu reviewed (commented):
Looking good - I also ran the benchmark tests and it doesn't seem to impact the performance which is nice,

On master,
```
=== RUN   TestLightningNetworkDaemon/tranche00/117-of-275/btcd/async_payments_benchmark
    lnd_payment_test.go:757: 	Benchmark info: Elapsed time:  40.501505292s
    lnd_payment_test.go:758: 	Benchmark info: TPS:  11.925482683119037

=== RUN   TestLightningNetworkDaemon/tranche00/225-of-275/btcd/remote_signer-funding_async_payments
    lnd_payment_test.go:757: 	Benchmark info: Elapsed time:  42.424433709s
    lnd_payment_test.go:758: 	Benchmark info: TPS:  11.384948666917278


=== RUN   TestLightningNetworkDaemon/tranche00/226-of-275/btcd/remote_signer-funding_async_payments_taproot
    lnd_payment_test.go:757: 	Benchmark info: Elapsed time:  42.670583291s
    lnd_payment_test.go:758: 	Benchmark info: TPS:  11.319273437302027
```


This branch,
```
=== RUN   TestLightningNetworkDaemon/tranche00/117-of-274/btcd/async_payments_benchmark
    lnd_payment_test.go:757:    Benchmark info: Elapsed time:  43.210928459s
    lnd_payment_test.go:758:    Benchmark info: TPS:  11.177727885627055

=== RUN   TestLightningNetworkDaemon/tranche00/224-of-274/btcd/remote_signer-funding_async_payments
    lnd_payment_test.go:757:    Benchmark info: Elapsed time:  43.428684625s
    lnd_payment_test.go:758:    Benchmark info: TPS:  11.121681537689446

=== RUN   TestLightningNetworkDaemon/tranche00/225-of-274/btcd/remote_signer-funding_async_payments_taproot
    lnd_payment_test.go:757:    Benchmark info: Elapsed time:  46.923330958s
    lnd_payment_test.go:758:    Benchmark info: TPS:  10.293386896005364
```

Think once the CI is fixed we are good to go.

Roasbeef commented:
> I also ran the benchmark tests and it doesn't seem to impact the performance which is nice,

That's expected, since this should only affect gossip syncers. We don't rate limit our own commit sig or HTLC messages. 

What do we think about the set of defaults? Is 100 KB/s reasonable for most nodes?

Roasbeef commented:
Fixed a bug (was returning early from the send loop). 

Also modified the structure to allow us to cancel a rate limit reservation if a peer exits. This way we release those tokens/bytes back to the bucket.

yyforyongyu reviewed (commented):
Close! Some typos and my main comment is whether we should add a simple validation for the user configs, to make sure we won't be blocked on the delay forever.

> Also modified the structure to allow us to cancel a rate limit reservation if a peer exits. This way we release those tokens/bytes back to the bucket.

If the peer reconnects I think a new rate limiter will be created with a bucket full of tokens.

Roasbeef commented:
> If the peer reconnects I think a new rate limiter will be created with a bucket full of tokens.

So there's a global bucket. The change I made handles the case where a peer sleeps to send 65 KB, but then disconnects before they actually send (or disconnects mid sleep). We'll now release that reservation back into the available pool.

Roasbeef commented on discovery/sync_manager.go:
```diff
@@ -494,6 +543,91 @@ func (m *SyncManager) isPinnedSyncer(s *GossipSyncer) bool {
 
+// deriveRateLimitReservation will take the current message and derive a
+// reservation that can be used to wait on the rate limiter.
+func (m *SyncManager) deriveRateLimitReservation(ctx context.Context,
```
Ah, left over after the refactor, removed.

Roasbeef commented on discovery/sync_manager_test.go:
```diff
@@ -685,3 +689,177 @@ func assertActiveSyncerTransition(t *testing.T, s *GossipSyncer, peer *mockPeer)
+func TestDeriveRateLimitReservation(t *testing.T) {
+	// Define standard test parameters for rate limiting.
+	bytesPerSec := uint64(1000) // 1000 bytes/second
+	bytesBurst := uint64(100)   // 100 bytes burst capacity
```
I think this is ok for the test, as we're just making assertions based on the delay value, rather than having the test sleep. In the future I think something like `[synctest](https://go.dev/blog/synctest)` would allow us to actually sleep here to assert messages are _eventually_ sent.

Roasbeef commented on lncfg/gossip.go:
```diff
@@ -32,6 +32,10 @@ type Gossip struct {
 
 	AnnouncementConf uint32 `long:"announcement-conf" description:"The number of confirmations required before processing channel announcements."`
+
+	MsgRateBytes uint64 `long:"msg-rate-bytes" description:"The maximum number of bytes of gossip messages that will be sent per second. This is a global limit that applies to all peers."`
```
SGTM.

Roasbeef commented on discovery/sync_manager.go:
```diff
@@ -494,6 +543,91 @@ func (m *SyncManager) isPinnedSyncer(s *GossipSyncer) bool {
+	// If we've already replied a handful of times, we will start to delay
+	// responses back to the remote peer. This can help prevent DOS attacks
+	// where the remote peer spams us endlessly.
+	delay := limitReservation.Delay()
```
Added a comment, also the validation elsewhere.

Crypt-iQ reviewed (commented):
Concept ACK. Reviewed the code and I think this is a good change. I think this needs to be tested under real load during RC to figure out better values. My only concern is that depending on the values, it might be possible that an attacker can get honest peers rate-limited by asking for a graph dump since the limiter is shared. So I think good values here will go a long way. Alternatively, maybe we could note down all the scenarios where peers will request routing dumps and then evaluate if all of those cases are necessary?

Roasbeef commented:
> My only concern is that depending on the values, it might be possible that an attacker can get honest peers rate-limited by asking for a graph dump since the limiter is shared

FWIW, we still have that global semaphore to limit the number of outstanding gossip filter requests. Moving to a per peer limiter is possible, then then an attacker can just spin up multiple peers to get around it. 

The original motivation for this change was to move to a global rate limiting value, as in the wild for certain degenerate cases, all the peers were below the rate limit, but in aggregate (hundreds of peers) bandwidth utilization was very high.

ziggie1984 commented on discovery/sync_manager.go:
```diff
@@ -180,8 +212,25 @@ func newSyncManager(cfg *SyncManagerCfg) *SyncManager {
+		bytesPerSecond = DefaultMsgBytesPerSecond
+	}
+
+	bytesBurst := cfg.AllotedMsgBytesBurst
```
Should we do some boundary checks for these constants ?

Roasbeef replied:
I added a check in the config validation. We need this value to be above 65 KB.

ziggie1984 commented on discovery/sync_manager.go:
```diff
@@ -494,6 +543,95 @@ func (m *SyncManager) isPinnedSyncer(s *GossipSyncer) bool {
+	peer lnpeer.Peer, nodeID route.Vertex, msgs ...lnwire.Message) error {
+
+	for _, msg := range msgs {
+		if err := m.maybeRateLimitMsg(ctx, nodeID, msg); err != nil {
```
I wonder if we should skip rate-limiting when `sync=true`, imo we should not really rate-limit if we require a peer synchronization.

Roasbeef replied:
IIUC `sync=true` here just means we should block until the message has been fully sent out.

ziggie1984 commented on discovery/sync_manager_test.go:
```diff
@@ -685,3 +688,167 @@ func assertActiveSyncerTransition(t *testing.T, s *GossipSyncer, peer *mockPeer)
 }
+
+// TestSizeableMessage is a test implementation of lnwire.SizeableMessage.
+type TestSizeableMessage struct {
```
why did you not use a mock ?

Roasbeef replied:
Just needed some simpler behavior so I didn't reach for it 🤷

ziggie1984 commented on discovery/sync_manager_test.go:
```diff
@@ -685,3 +688,167 @@ func assertActiveSyncerTransition(t *testing.T, s *GossipSyncer, peer *mockPeer)
+	})
+
+	// Test that message size affects delay.
+	t.Run("larger messages have longer delays", func(t *testing.T) {
```
Nit: seems like we are testing the rateLimiter here, I think we might not really need this testcase ?

Roasbeef replied:
In a sense that's true, but at least we're able to bind the current implementation to a basic property based assumption. 

A larger test would need to actually sleep to make sure we're able to eventually send messages. Once we can use `synctest`, we can add something like that.

ziggie1984 commented on sample-lnd.conf:
```diff
@@ -1743,6 +1743,17 @@
 ; The number of confirmations required before processing channel announcements.
 ; gossip.announcement-conf=6
 
+; The allotted bandwdith rate expressed in bytes/second that will be alloacted
```
Nit: alloacted => allocated, bandwdith => bandwidth

ziggie1984 reviewed (approved):
LGTM, really easy to review change 👌
